modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/GraphSpec.scala

```diff
@@ -0,0 +1,143 @@
+package ch.epfl.bluebrain.nexus.rdf
+
+import cats.kernel.Eq
+import cats.syntax.show._
+import ch.epfl.bluebrain.nexus.rdf.Graph.Triple
+import ch.epfl.bluebrain.nexus.rdf.Node.{IriNode, IriOrBNode, Literal}
+import ch.epfl.bluebrain.nexus.rdf.syntax.node._
+import ch.epfl.bluebrain.nexus.rdf.syntax.node.unsafe._
+import org.scalatest.{EitherValues, Matchers, WordSpecLike}
+
+import scala.collection.immutable.HashSet
+
+//noinspection TypeAnnotation
+class GraphSpec extends WordSpecLike with Matchers with EitherValues {
+
+  "An RDF Graph" should {
+    val a      = url"http://a"
+    val isa    = url"http://isa"
+    val hasa   = url"http://hasa"
+    val string = url"http://string"
+    val bool   = url"http://bool"
+
+    object p {
+      val prefix = "http://prop"
+      val string = url"$prefix/string"
+      val int    = url"$prefix/int"
+      val bool   = url"$prefix/bool"
+      val other  = url"$prefix/other"
+    }
+
+    val triples = Set[(IriOrBNode, IriNode, Node)](
+      (a, isa, string),
+      (a, isa, bool),
+      (a, hasa, b"1"),
+      (b"1", p.string, "asd"),
+      (b"1", p.int, 2),
+      (b"1", p.bool, true)
+    )
+
+    val g = Graph(triples)
+
+    "return the same collection of triples" in {
+      g.triples shouldEqual triples
+    }
+
+    "add a new triple" in {
+      val t: Triple = (b"1", p.int, 3)
+      (g + t).triples shouldEqual (triples + t)
+    }
+
+    "subtract a triple" in {
+      val t: Triple = (b"1", p.int, 2)
+      (g - t).triples shouldEqual (triples - t)
+    }
+
+    "subtract a missing triple" in {
+      val t: Triple = (b"1", p.int, 3)
+      (g - t).triples shouldEqual triples
+    }
+
+    "union with another graph" in {
+      val otherTriples = Set[Triple]((a, isa, string), (b"1", p.int, 4))
+      val g2           = Graph(otherTriples)
+      (g ++ g2).triples shouldEqual (triples + ((b"1", p.int, 4)))
+    }
+
+    "subtract a sub graph" in {
+      ((g + ((b"1", p.other, true))) -- g).triples shouldEqual Set[Triple]((b"1", p.other, true))
+    }
+
+    "subtract a disjoint graph" in {
+      val g2 = Graph((b"1", p.other, true): Triple)
+      (g -- g2).triples shouldEqual g.triples
+    }
+
+    "ignore adding existing triples" in {
+      (g + ((b"1", p.bool, true))).triples shouldEqual triples
+    }
+
+    "be cyclic" in {
+      val g = Graph((a, hasa, b"1"), (b"1", isa, string), (b"1", hasa, a))
```

Comment:
can you call it `g3` or something? Is a bit confusing when reading the test, even though it is correct because it has local scope